contracts/6.12/liquidator/PCSFlashLiquidator.sol

```diff
@@ -0,0 +1,117 @@
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+import "@alpaca-finance/alpaca-contract/contracts/6/protocol/apis/pancake/IPancakeRouter02.sol";
+
+import "../interfaces/IFlashLendingCallee.sol";
+import "../interfaces/IGenericTokenAdapter.sol";
+import "../interfaces/IBookKeeper.sol";
+import "../interfaces/IAlpacaVault.sol";
+import "../interfaces/IStableSwapModule.sol";
+import "../interfaces/IStablecoinAdapter.sol";
+import "../utils/SafeToken.sol";
+
+contract PCSFlashLiquidator is OwnableUpgradeable, IFlashLendingCallee {
+  using SafeToken for address;
+  using SafeMathUpgradeable for uint256;
+
+  // --- Math ---
+  uint256 constant WAD = 10**18;
+  uint256 constant RAY = 10**27;
+  uint256 constant RAD = 10**45;
+
+  IBookKeeper public bookKeeper;
+  IStablecoinAdapter public stablecoinAdapter;
+  address public alpacaStablecoin;
+
+  function initialize(
+    address _bookKeeper,
+    address _alpacaStablecoin,
+    address _stablecoinAdapter
+  ) external initializer {
+    OwnableUpgradeable.__Ownable_init();
+
+    bookKeeper = IBookKeeper(_bookKeeper);
+    alpacaStablecoin = _alpacaStablecoin;
+    stablecoinAdapter = IStablecoinAdapter(_stablecoinAdapter);
+  }
+
+  function flashLendingCall(
+    address _caller,
+    uint256 _debtValueToRepay, // [rad]
+    uint256 _collateralAmountToLiquidate, // [wad]
+    bytes calldata data
+  ) external override {
+    (
+      address _liquidatorAddress,
+      IGenericTokenAdapter _tokenAdapter,
+      address _vaultAddress,
+      IPancakeRouter02 _router,
+      address[] memory _path
+    ) = abi.decode(data, (address, IGenericTokenAdapter, address, IPancakeRouter02, address[]));
+
+    // Retrieve collateral token
+    (address _token, uint256 _actualCollateralAmount) = _retrieveCollateral(
+      _tokenAdapter,
+      _vaultAddress,
+      _collateralAmountToLiquidate
+    );
+
+    // Swap token to AUSD
+    require(
+      _debtValueToRepay.div(RAY) <= _sellCollateral(_token, _path, _router, _actualCollateralAmount, _debtValueToRepay),
+      "not enough to repay debt"
+    );
+
+    // Deposit Alpaca Stablecoin for liquidatorAddress
+    _depositAlpacaStablecoin(_debtValueToRepay.div(RAY), _liquidatorAddress);
+  }
+
+  function _retrieveCollateral(
+    IGenericTokenAdapter _tokenAdapter,
+    address _vaultAddress,
+    uint256 _amount
+  ) internal returns (address _token, uint256 _actualAmount) {
+    bookKeeper.whitelist(address(_tokenAdapter));
```

Comment:
Should this be just one time thing?

Reply:
Different TokenAdapter could be passed down to this function, because different collateral pools have different adapters. This needs to be there for the `TokenAdapter.withdraw()` to work. It could be prettier by checking if we have already whitelisted this TokenAdapter. But I think it would waste more gas. Are you okay if I don't change anything here?

Reply:
Sounds ok to me.